joinmarket-webui/jam
# feat: ability to retrieve logs

Closes #469.

First iteration of "feat: ability to retrieve logs #469".
This PR uses features implemented in https://github.com/joinmarket-webui/jam-docker/pull/51 and must be tested in conjunction with it.

Features:
- ability to retrieve logs
- refresh log file contents
- auto scroll to bottom on refresh
- ~~colorize lines by log level~~ (reverted: uses too much memory)

Possible improvements for the next iteration:
- auto refresh (aka "follow log output")
- filter by log level
- colorize lines by log level

---

Copying a phrase from the other PR:
> As always, inputs are highly appreciated and especially welcomed on this PR, as it contains rather impactful changes.
If you have any idea how to do things better or in a smarter way, please let me know!

Also, if you have any questions or are unsure about anything - don't hold back! Happy to answer and discuss any upcoming doubts/thoughts/ideas. 



## **OLD** How to test (outdated: see below)
1. Rebuild the docker image for the secondary regtest container
  - Checkout branch [`serve-logs`](url) from `jam-docker`. See PR https://github.com/joinmarket-webui/jam-docker/pull/51 for more details.
  - Rebuild with
  ```
docker build --label "local" \
        --build-arg JAM_REPO_REF=master \
        --build-arg JM_SERVER_REPO_REF=master \
        --tag "joinmarket-webui/jam-standalone-logs-local" ./standalone
```
2. Change the secondary docker image to use the newly built image
```patch
diff --git a/docker/regtest/dockerfile-deps/joinmarket/webui-standalone/Dockerfile b/docker/regtest/dockerfile-deps/joinmarket/webui-standalone/Dockerfile
index 84a1e68..f8a8ec4 100644
--- a/docker/regtest/dockerfile-deps/joinmarket/webui-standalone/Dockerfile
+++ b/docker/regtest/dockerfile-deps/joinmarket/webui-standalone/Dockerfile
@@ -1,4 +1,5 @@
-FROM ghcr.io/joinmarket-webui/jam-dev-standalone:master
+#FROM ghcr.io/joinmarket-webui/jam-dev-standalone:master
+FROM joinmarket-webui/jam-standalone-logs-local 
```
3. Rebuild your local regtest setup, do not use `npm run regtest:rebuild`, but:
```sh
npm run regtest:clear && docker-compose \
        --env-file docker/regtest/.env.example \
        --file docker/regtest/docker-compose.yml build
```
(why? -> because `--pull` is not supported if you have a locally built image in your docker-compose setup)


4. start the regtest env as normal, but instead of `npm start`, run:
```
npm run start:secondary
```

## How to test (up-to-date)
1. Rebuild your regtest env `npm run regtest:rebuild`
2. Start Jam with `npm run start:regtest:secondary` 

## :camera_flash: 
<img src="https://user-images.githubusercontent.com/3358649/186024238-ab6658f3-67d3-4c1d-9045-bc28e6cf99d8.png" width=400 />

<img src="https://user-images.githubusercontent.com/3358649/186042283-d688a153-778d-4fbe-8df3-f08ad48728a3.png" width=400 /><img src="https://user-images.githubusercontent.com/3358649/186042298-97a6d7ab-795f-4caa-8ea6-33f3bcce0d46.png" width=400 />

##### :camera_flash:  reverted changes (just for reference; might re-implement in future iterations) 
<img src="https://user-images.githubusercontent.com/3358649/186023502-d0289255-d1be-49ae-856d-d7106ee050aa.png" width=400 /><img src="https://user-images.githubusercontent.com/3358649/186023473-6450a7fd-d511-493e-9b34-9b6ff6a80dc8.png" width=400 />
